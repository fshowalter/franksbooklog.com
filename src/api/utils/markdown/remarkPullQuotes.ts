import type { Root } from "mdast";
import type { Plugin } from "unified";

import { visit } from "unist-util-visit";

export const remarkPullQuotes: Plugin<[], Root> = () => {
  return (tree: Root) => {
    visit(tree, "blockquote", (node) => {
      const firstChild = node.children[0];

      if (firstChild && firstChild.type === "paragraph") {
        const firstText = firstChild.children[0];

        if (
          firstText &&
          firstText.type === "text" &&
          firstText.value.startsWith("{pullquote}")
        ) {
          // Remove the marker from the text
          firstText.value = firstText.value.replace("{pullquote}", "").trim();

          // Add pull-quote class to the blockquote
          node.data = {
            ...node.data,
            hProperties: {
              ...(node.data?.hProperties as Record<string, unknown>),
              className: "pull-quote",
            },
          };
        }
      }
    });
  };
};
