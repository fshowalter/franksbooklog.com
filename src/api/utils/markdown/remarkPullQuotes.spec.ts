import rehypeStringify from "rehype-stringify";
import { remark } from "remark";
import remarkRehype from "remark-rehype";
import { describe, it } from "vitest";

import { remarkPullQuotes } from "./remarkPullQuotes";

function processMarkdown(markdown: string): string {
  return remark()
    .use(remarkPullQuotes)
    .use(remarkRehype)
    .use(rehypeStringify)
    .processSync(markdown)
    .toString();
}

describe("remarkPullQuotes", () => {
  it("converts {pullquote} marker to pull-quote class", ({ expect }) => {
    const markdown = `> {pullquote} This is a memorable quote from the book.`;
    const result = processMarkdown(markdown);

    expect(result).toMatchSnapshot();
  });

  it("removes the marker but preserves quote content", ({ expect }) => {
    const markdown = `> {pullquote} An organization does well only those things the Boss checks.`;
    const result = processMarkdown(markdown);

    expect(result).toMatchSnapshot();
  });

  it("leaves regular blockquotes unchanged", ({ expect }) => {
    const markdown = `> This is a regular blockquote without the marker.`;
    const result = processMarkdown(markdown);

    expect(result).toMatchSnapshot();
  });

  it("handles empty pullquote marker", ({ expect }) => {
    const markdown = `> {pullquote}`;
    const result = processMarkdown(markdown);

    expect(result).toMatchSnapshot();
  });
});
