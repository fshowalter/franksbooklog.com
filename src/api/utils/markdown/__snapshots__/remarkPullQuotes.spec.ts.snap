// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`remarkPullQuotes > converts {pullquote} marker to pull-quote class 1`] = `
"<blockquote class="pull-quote">
<p>This is a memorable quote from the book.</p>
</blockquote>"
`;

exports[`remarkPullQuotes > handles empty pullquote marker 1`] = `
"<blockquote class="pull-quote">
<p></p>
</blockquote>"
`;

exports[`remarkPullQuotes > leaves regular blockquotes unchanged 1`] = `
"<blockquote>
<p>This is a regular blockquote without the marker.</p>
</blockquote>"
`;

exports[`remarkPullQuotes > removes the marker but preserves quote content 1`] = `
"<blockquote class="pull-quote">
<p>An organization does well only those things the Boss checks.</p>
</blockquote>"
`;
